tesseract/tesseract_planning/tesseract_process_managers/include/tesseract_process_managers/process_planning_server.h

```diff
@@ -0,0 +1,274 @@
+
+  /** @brief If the environment has changed it will rebuild the cache of tesseract objects */
+  virtual void refreshCache();
+
+  /**
+   * @brief This will pop a Tesseract object from the queue
+   * @details This will first call refreshCache to ensure it has an updated tesseract then proceed
+   */
+  virtual tesseract::Tesseract::Ptr getCachedTesseract();
+
+protected:
+  /** @brief The tesseract_object used to create the cache */
+  tesseract::Tesseract::Ptr tesseract_;
+
+  /** @brief The environment revision number at the time the cache was populated */
+  int cache_env_revision_{ 0 };
+
+  /** @brief The assigned cache size */
+  std::size_t cache_size_{ 5 };
+
+  /** @brief A vector of cached Tesseact objects */
+  std::deque<tesseract::Tesseract::Ptr> cache_;
+
+  /** @brief The mutex used when reading and writing to cache_ */
+  mutable std::shared_mutex cache_mutex_;
+};
+
+struct ProcessPlanningRequest
+{
+  /** @brief The name of the planner to use */
+  std::string name;
+
+  /** @brief  This should an xml string of the command language instructions */
+  Instruction instructions{ NullInstruction() };
+
+  /** @brief This should an xml string of the command language instructions (Optional) */
+  Instruction seed{ NullInstruction() };
+
+  /** @brief Environment state to start planning with (Optional)  */
+  tesseract_environment::EnvState::ConstPtr env_state;
+
+  /** @brief Additional Commands to be applied to environment prior to planning (Optional) */
+  tesseract_environment::Commands commands;
+
+  /** @brief Enable debug content (Optional) */
+  bool debug{ false };
+
+  /** @brief Enable profiling of the planning request (Optional) */
+  bool profile{ false };
+
+  /**
+   * @brief This allows the remapping of the Plan Profile identified in the command language to a specific profile for a
+   * given motion planner. (Optional)
+   */
+  PlannerProfileRemapping plan_profile_remapping;
+
+  /**
+   * @brief This allows the remapping of the Composite Profile identified in the command language to a specific profile
+   * for a given motion planner. (Optional)
+   */
+  PlannerProfileRemapping composite_profile_remapping;
+};
+
+/**
+ * @brief This contains the result for the process planning request
+ * @details Must chec the status before access the results to know if available.
+ * @note This must not go out of scope until the process has finished
+ */
+struct ProcessPlanningFuture
+{
+  std::future<void> status;
+  std::unique_ptr<Instruction> input;
+  std::unique_ptr<Instruction> results;
+  std::unique_ptr<const ManipulatorInfo> global_manip_info;
+  std::unique_ptr<const PlannerProfileRemapping> plan_profile_remapping;
+  std::unique_ptr<const PlannerProfileRemapping> composite_profile_remapping;
+  TaskflowGenerator::UPtr taskflow_generator;
+};
+
+class ProcessPlanningServer
+{
+public:
+  ProcessPlanningServer(TesseractCache::Ptr cache, size_t n = std::thread::hardware_concurrency());
```

Comment:
Maybe a constructor that takes a Tesseract::Ptr as well?

Reply:
The TesseractCache requires a Tesseract::Ptr on construction. I think we may be able to remove the TesseractCache and use the one in your open PR but I have not taken a look yet.